### PR TITLE
add cost budget; move loc of make_dir

### PR DESCRIPTION
**Pull request by sonichi**
## Why are these changes needed?

1. Allow users to specify a cost budget when cost_attr is neither time nor num_samples in BlendSearch.
2. create a directory for log_file_name even when use_ray is true, to avoid FileNotFoundError.

## Checks

- [x] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR, or I've made sure [lint with flake8](https://github.com/microsoft/FLAML/blob/816a82a1155b4de4705b21a615ccdff67c6da379/.github/workflows/python-package.yml#L54-L59) output is two 0s.
- [ ] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

File: flaml/tune/searcher/blendsearch.py [hunk 1]
```diff
@@ -693,6 +700,8 @@ def on_trial_result(self, trial_id: str, result: Dict):
     def suggest(self, trial_id: str) -> Optional[Dict]:
         """choose thread, suggest a valid config."""
         if self._init_used and not self._points_to_evaluate:
+            if self._cost_budget and self._cost_used >= self._cost_budget:
+                return None
```
Comment (ZviBaratz) [inline flaml/tune/searcher/blendsearch.py:704]:
```suggestion
                return 
```

File: flaml/tune/searcher/blendsearch.py [hunk 2]
```diff
@@ -154,8 +157,10 @@ def __init__(
                 self.cost_attr = TIME_TOTAL_S
             else:
                 self.cost_attr = None
+            self._cost_budget = None
         else:
             self.cost_attr = cost_attr
+            self._cost_budget = cost_budget
```
Comment (liususan091219) [inline flaml/tune/searcher/blendsearch.py:163]: why is there _?

**Review by Yard1 — APPROVED** [review]
Looks fine to me from the Ray side

File: flaml/tune/searcher/blendsearch.py (hunk 2 above)
Comment (sonichi) [reply flaml/tune/searcher/blendsearch.py:163]: internal variable.

File: flaml/tune/searcher/blendsearch.py (hunk 1 above)
Comment (thinkall) [reply flaml/tune/searcher/blendsearch.py]: Why remove "None" from "return None"? `suggest` should return `config` if it ends sucessfully, I think `return None` here indicates that this function should return some value, just at this point, it returns None.

Considering `def suggest(self, trial_id: str) -> Optional[Dict]:`, maybe we can use `return {}`?

**Review by qingyun-wu — APPROVED** [review]

**Pull request merged**